File: apps/ari/frontend/src/api/audit.ts
/**
 * API client for the full audit pipeline endpoints.
 * Uses POST-based SSE via fetch + ReadableStream (same pattern as liteReport.ts).
 */

const API_BASE = '/api/v1'

// --- Data models ---

export interface FounderProfile {
  name: string
  title: string
  background: string
  prior_companies: string[]
  ai_name_collision_risk: boolean
}

export interface ProductInfo {
  name: string
  category: string
  description: string
  differentiators: string[]
}

export interface CompetitorInfo {
  name: string
  domain: string
}

export interface BrandProfile {
  company_name: string
  domain: string
  industry: string
  description: string
  entity_type: string
  competitors: CompetitorInfo[]
  personas: string[]
  topics: string[]
  differentiators: string[]
  legal_entity: string
  aliases: string[]
  founded: string
  headquarters: string
  founders: FounderProfile[]
  products: ProductInfo[]
  distribution_channels: string[]
  awards: string[]
  press_mentions: string[]
  brand_voice: string
  use_cases: string[]
  occasions: string[]
  regions: string[]
  adjacent_categories: string[]
  category_leader: string
  category_maturity: string
  sibling_brands: string[]
}

export interface DimensionScore {
  dimension: string
  score: number
  mention_rate: number
  prompt_count: number
  avg_position: number | null
}

export interface ProviderScore {
  provider: string
  score: number
  mention_rate: number
  prompt_count: number
  avg_position: number | null
}

export interface ScoringData {
  overall_ari: number
  severity_band: string
  mention_frequency: number
  position_quality: number
  narrative_accuracy: number
  founder_retrieval: number
}

export interface AntiPattern {
  pattern_type: string
  display_name: string
  severity: string
  evidence: string
  recommendation: string
}

export interface GapAnalysis {
  gap_type: string
  description: string
  impact: number
  effort: number
  coverage: number
  priority_score: number
  recommendation: string
}

export interface AuditReport {
  executive_summary: string
  core_problem: string
  core_problem_name: string
  competitive_landscape: string
  dimension_analysis: string
  anti_patterns_section: string
  recommendations: string
  pitch_hook: string
  full_markdown: string
}

// --- SSE event types ---

export interface DomainValidation {
  domain: string
  title: string
  meta_description: string
  url: string
}

export interface DiscoveryResult {
  company_name: string
  domain: string
  industry: string
  entity_type: string
  description: string
  competitors: CompetitorInfo[]
  personas: string[]
  topics: string[]
  differentiators: string[]
}

export interface AuditPromptStart {
  type: 'prompt_start'
  current: number
  total: number
  provider: string
  dimension: string
  persona: string
  topic: string
  prompt_text: string
}

export interface AuditPromptResult {
  type: 'prompt_result'
  current: number
  total: number
  provider: string
  dimension: string
  persona: string
  topic: string
  mentioned: boolean
  position: number | null
}

export type AuditEvent =
  | { type: 'status'; status: string; message: string }
  | { type: 'domain_validated'; data: DomainValidation }
  | { type: 'discovery_complete'; data: DiscoveryResult }
  | { type: 'profile_complete'; data: BrandProfile }
  | { type: 'matrix_complete'; data: { total_prompts: number; dimensions: string[] } }
  | AuditPromptStart
  | AuditPromptResult
  | { type: 'analysis_complete'; data: { total_results: number; mentions: number } }
  | { type: 'scoring_complete'; data: ScoringData }
  | { type: 'anti_patterns_complete'; data: { anti_patterns: AntiPattern[]; gaps: GapAnalysis[] } }
  | { type: 'report_stage'; stage: string }
  | { type: 'report_complete'; data: AuditReport }
  | { type: 'pdf_ready'; job_id: string; status: string }
  | { type: 'error'; status: string; message: string }

// --- API functions ---

/**
 * Start the full audit pipeline via POST-based SSE.
 * Returns a cleanup function to abort the request.
 */
export function startAuditAnalysis(
  domain: string,
  onEvent: (event: AuditEvent) => void,
  onError: (error: Error) => void,
  options?: {
    brandProfileOverride?: Record<string, unknown>
    liteDiscovery?: DiscoveryResult
    forceRerun?: boolean
  },
): () => void {
  const controller = new AbortController()

  const body: Record<string, unknown> = { domain }
  if (options?.brandProfileOverride) body.brand_profile_override = options.brandProfileOverride
  if (options?.liteDiscovery) body.lite_discovery = options.liteDiscovery
  if (options?.forceRerun) body.force_rerun = true

  fetch(`${API_BASE}/audit/analyze`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal: controller.signal,
  })
    .then(async (response) => {
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`)
      }

      const reader = response.body?.getReader()
      if (!reader) throw new Error('No response body')

      const decoder = new TextDecoder()
      let buffer = ''

      while (true) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })

        const lines = buffer.split('\n')
        buffer = lines.pop() || ''

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            try {
              const data = JSON.parse(line.slice(6))
              onEvent(data as AuditEvent)
            } catch {
              // Skip malformed lines
            }
          }
        }
      }
    })
    .catch((err) => {
      if (err.name !== 'AbortError') {
        onError(err)
      }
    })

  return () => controller.abort()
}

/**
 * Run just the brand profiling phase.
 */
export async function profileOnly(domain: string): Promise<BrandProfile> {
  const response = await fetch(`${API_BASE}/audit/profile-only`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ domain }),
  })

  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: 'Unknown error' }))
    throw new Error(err.detail || `HTTP ${response.status}`)
  }

  return response.json()
}

/**
 * Get the download URL for a generated audit PDF.
 */
export function getAuditDownloadUrl(jobId: string): string {
  return `${API_BASE}/audit/download/${jobId}`
}

File: apps/ari/backend/app/models/audit.py
"""Data models for the full audit pipeline (Fancy Robot methodology)."""

from datetime import datetime
from enum import Enum
from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from app.models.lite_report import CompetitorInfo, DiscoveryResult


# --- Enums ---


class SeverityBand(str, Enum):
    """Severity classification based on overall ARI score."""

    CRITICAL = "critical"        # 0-15
    POOR = "poor"                # 16-30
    BELOW_AVERAGE = "below_avg"  # 31-45
    MODERATE = "moderate"        # 46-60
    GOOD = "good"                # 61-75
    STRONG = "strong"            # 76-90
    DOMINANT = "dominant"        # 91-100


class AuditPromptDimension(str, Enum):
    """The 8 prompt dimensions from the Fancy Robot methodology."""

    CATEGORY_DEFAULT = "category_default"
    USE_CASE = "use_case"
    COMPARISON = "comparison"
    ATTRIBUTE_SPECIFIC = "attribute_specific"
    GIFT_SOCIAL = "gift_social"
    FOUNDER_BRAND = "founder_brand"
    GEOGRAPHIC = "geographic"
    ADJACENT_CATEGORY = "adjacent_category"


class AntiPatternType(str, Enum):
    """The 10 named anti-patterns from the spec."""

    KLEENEX_EFFECT = "kleenex_effect"
    PREMIUM_TAX = "premium_tax"
    MESSY_MIDDLE = "messy_middle"
    FOUNDER_INVISIBILITY = "founder_invisibility"
    AWARD_AMNESIA = "award_amnesia"
    NAME_FRAGMENTATION = "name_fragmentation"
    DUAL_CATEGORY_TRAP = "dual_category_trap"
    SOCIAL_PROOF_NOT_AI_PROOF = "social_proof_not_ai_proof"
    PORTFOLIO_ISOLATION = "portfolio_isolation"
    AFFILIATE_DISTORTION = "affiliate_distortion"


class GapType(str, Enum):
    """Types of visibility gaps."""

    CONTENT = "content"
    NARRATIVE = "narrative"
    COMPETITIVE = "competitive"
    STRUCTURAL = "structural"
    CATEGORY = "category"


# --- Brand Profile (superset of DiscoveryResult) ---


class FounderProfile(BaseModel):
    """Detailed founder/leader profile."""

    name: str
    title: str = ""
    background: str = ""
    prior_companies: list[str] = Field(default_factory=list)
    ai_name_collision_risk: bool = False


class ProductInfo(BaseModel):
    """A product or service line."""

    name: str
    category: str = ""
    description: str = ""
    differentiators: list[str] = Field(default_factory=list)


class BrandProfile(BaseModel):
    """Deep brand profile - superset of DiscoveryResult with audit-specific fields."""

    # Core (from discovery)
    company_name: str
    domain: str
    industry: str = ""
    description: str = ""
    entity_type: str = "company"
    competitors: list[CompetitorInfo] = Field(default_factory=list)
    personas: list[str] = Field(default_factory=list)
    topics: list[str] = Field(default_factory=list)
    differentiators: list[str] = Field(default_factory=list)

    # Extended (audit-specific)
    legal_entity: str = ""
    aliases: list[str] = Field(default_factory=list)
    founded: str = ""
    headquarters: str = ""
    founders: list[FounderProfile] = Field(default_factory=list)
    products: list[ProductInfo] = Field(default_factory=list)
    distribution_channels: list[str] = Field(default_factory=list)
    awards: list[str] = Field(default_factory=list)
    press_mentions: list[str] = Field(default_factory=list)
    brand_voice: str = ""
    use_cases: list[str] = Field(default_factory=list)
    occasions: list[str] = Field(default_factory=list)
    regions: list[str] = Field(default_factory=list)
    adjacent_categories: list[str] = Field(default_factory=list)
    category_leader: str = ""  # The "Kleenex brand" for this category
    category_maturity: str = ""  # emerging, growing, mature, saturated
    sibling_brands: list[str] = Field(default_factory=list)

    @classmethod
    def from_discovery(cls, discovery: DiscoveryResult) -> "BrandProfile":
        """Upgrade a DiscoveryResult to a BrandProfile."""
        return cls(
            company_name=discovery.company_name,
            domain=discovery.domain,
            industry=discovery.industry,
            description=discovery.description,
            entity_type=discovery.entity_type,
            competitors=discovery.competitors,
            personas=discovery.personas,
            topics=discovery.topics,
            differentiators=discovery.differentiators,
        )


# --- Prompt Models ---


class AuditRenderedPrompt(BaseModel):
    """A single rendered prompt for the audit pipeline."""

    id: str
    text: str
    dimension: AuditPromptDimension
    persona: str = ""
    topic: str = ""
    weight: float = 1.0
    competitor: str = ""
    metadata: dict = Field(default_factory=dict)


# --- Analysis Results ---


class PromptAnalysisResult(BaseModel):
    """Result of a single prompt across one provider."""

    prompt_id: str
    prompt_text: str
    dimension: AuditPromptDimension
    persona: str = ""
    topic: str = ""
    provider: str
    model_version: str = ""
    raw_response: str = ""
    brand_mentioned: bool = False
    position: int | None = None
    recommendation_type: str = "not_mentioned"
    sentiment: str = "neutral"
    confidence: float = 0.0
    context: str = ""
    latency_ms: int = 0
    tokens_used: int | None = None


class DimensionScore(BaseModel):
    """Score breakdown for a single prompt dimension."""

    dimension: AuditPromptDimension
    score: float = 0.0
    mention_rate: float = 0.0
    prompt_count: int = 0
    avg_position: float | None = None


class ProviderAuditScore(BaseModel):
    """Score breakdown for a single provider."""

    provider: str
    score: float = 0.0
    mention_rate: float = 0.0
    prompt_count: int = 0
    avg_position: float | None = None


class AuditAnalysisResult(BaseModel):
    """Complete audit analysis result with 4-factor scoring."""

    # Overall
    overall_ari: float = 0.0
    severity_band: SeverityBand = SeverityBand.CRITICAL

    # 4-factor scores (each 0-100)
    mention_frequency: float = 0.0    # 40% weight
    position_quality: float = 0.0     # 25% weight
    narrative_accuracy: float = 0.0   # 20% weight
    founder_retrieval: float = 0.0    # 15% weight

    # Breakdowns
    dimension_scores: list[DimensionScore] = Field(default_factory=list)
    provider_scores: list[ProviderAuditScore] = Field(default_factory=list)
    persona_breakdown: dict[str, float] = Field(default_factory=dict)
    topic_breakdown: dict[str, float] = Field(default_factory=dict)
    competitor_mention_rates: dict[str, float] = Field(default_factory=dict)

    # Raw data
    total_prompts: int = 0
    total_responses: int = 0
    mentions_count: int = 0
    all_results: list[PromptAnalysisResult] = Field(default_factory=list)

    # Cost tracking
    total_tokens: int = 0
    estimated_cost_usd: float = 0.0


# --- Anti-Patterns ---


class DetectedAntiPattern(BaseModel):
    """A detected anti-pattern with evidence."""

    pattern_type: AntiPatternType
    display_name: str
    severity: str = "medium"  # low, medium, high, critical
    evidence: str = ""
    recommendation: str = ""


class GapAnalysis(BaseModel):
    """A visibility gap with priority scoring."""

    gap_type: GapType
    description: str
    impact: float = 0.0       # 0-1
    effort: float = 0.5       # 0-1 (lower = easier)
    coverage: float = 0.0     # 0-1 (how much of the problem this addresses)
    priority_score: float = 0.0  # impact * coverage * (1/effort)
    recommendation: str = ""


# --- Report ---


class AuditReport(BaseModel):
    """The final audit report with all narrative sections."""

    executive_summary: str = ""
    core_problem: str = ""
    core_problem_name: str = ""  # e.g. "The Premium Tax"
    competitive_landscape: str = ""
    dimension_analysis: str = ""
    anti_patterns_section: str = ""
    recommendations: str = ""
    pitch_hook: str = ""
    full_markdown: str = ""


# --- Top-level Audit Run ---


class AuditRunStatus(str, Enum):
    """Status progression for an audit run."""

    PENDING = "pending"
    PROFILING = "profiling"
    GENERATING_MATRIX = "generating_matrix"
    ANALYZING = "analyzing"
    SCORING = "scoring"
    DETECTING_PATTERNS = "detecting_patterns"
    COMPOSING_REPORT = "composing_report"
    GENERATING_PDF = "generating_pdf"
    COMPLETED = "completed"
    FAILED = "failed"


class AuditRun(BaseModel):
    """A complete audit run record."""

    id: UUID = Field(default_factory=uuid4)
    domain: str
    company_name: str = ""
    status: AuditRunStatus = AuditRunStatus.PENDING
    report_type: str = "full_audit"

    # Results
    brand_profile: BrandProfile | None = None
    analysis_result: AuditAnalysisResult | None = None
    anti_patterns: list[DetectedAntiPattern] = Field(default_factory=list)
    gap_analysis: list[GapAnalysis] = Field(default_factory=list)
    report: AuditReport | None = None
    pdf_url: str = ""

    # Denormalized
    overall_score: float = 0.0
    severity_band: SeverityBand = SeverityBand.CRITICAL

    # Cost
    cost_usd: float = 0.0

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: datetime | None = None

    # GFT link
    gft_company_id: str | None = None


# --- Utility ---


def get_severity_band(score: float) -> SeverityBand:
    """Map a 0-100 score to a severity band."""
    if score <= 15:
        return SeverityBand.CRITICAL
    elif score <= 30:
        return SeverityBand.POOR
    elif score <= 45:
        return SeverityBand.BELOW_AVERAGE
    elif score <= 60:
        return SeverityBand.MODERATE
    elif score <= 75:
        return SeverityBand.GOOD
    elif score <= 90:
        return SeverityBand.STRONG
    else:
        return SeverityBand.DOMINANT


SEVERITY_DISPLAY = {
    SeverityBand.CRITICAL: "Critical (0-15)",
    SeverityBand.POOR: "Poor (16-30)",
    SeverityBand.BELOW_AVERAGE: "Below Average (31-45)",
    SeverityBand.MODERATE: "Moderate (46-60)",
    SeverityBand.GOOD: "Good (61-75)",
    SeverityBand.STRONG: "Strong (76-90)",
    SeverityBand.DOMINANT: "Dominant (91-100)",
}
